File: Leetcode/563_Binary_Tree_Tilt.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <cmath>

class Solution {
public:
    int findTilt(TreeNode* root) {
        if(root == nullptr)
        {
            return 0;
        }
        
        if(root->left == nullptr && root->right == nullptr)
        {
            return 0;
        }
        
        if(root->left == nullptr)
        {
            return findTilt(root->right) + abs(findSum(root->right));
        }
        
        if(root->right == nullptr)
        {
            return findTilt(root->left) + abs(findSum(root->left));
        }
        
        int left = findSum(root->left);
        int right = findSum(root->right);
        
        return abs(left - right) + findTilt(root->right) + findTilt(root->left);
    }
    
    int findSum(TreeNode* root)
    {
        if(root == nullptr)
        {
            return 0;
        }
        
        return root->val + findSum(root->left) + findSum(root->right);
    }
};

File: Leetcode/563_Binary_Tree_Tilt.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findTilt = function (root) {
    tilt(root, 0);
    return traverseAndGetSum(root, 0);
};

var tilt = function (root, sum) {
    if (root) {
        const leftSum = tilt(root.left, sum);
        const rightSum = tilt(root.right, sum);


        const originalRootVal = root.val;
        root.val = Math.abs(leftSum - rightSum);

        return originalRootVal + leftSum + rightSum;
    }
    return 0;
}

var traverseAndGetSum = function (root, sum) {
    if (root) {
        const leftSum = traverseAndGetSum(root.left, sum);
        const rightSum = traverseAndGetSum(root.right, sum);

        return root.val + leftSum + rightSum;
    }
    return 0;
}